// State and enum values used for Modern Art, both for internal storage by the
// GameMaster, and in communication with the Players.

// The names of the artists, in order from least to most common.
message Artist {
  enum Id {
    LITE_METAL = 1;
    YOKO = 2;
    CHRISTIN_P = 3;
    KARL_GITTER = 4;
    KRYPTO = 5;
  }
}

// The different types of auction which may be associated with a painting/card.
message AuctionType {
  enum Id {
    SEALED = 1;
    FIXED = 2;
    ONCE_AROUND = 3;
    OPEN = 4;
    DOUBLE = 5;
  }
}

// A painting/card.
message Card {
  required Artist.Id artist = 1;
  required AuctionType.Id auction_type = 2;
}

// The state of a Player, as tracked by the GameMaster. (Players themselves are
// not treated as trustworthy, so critical data for each Player is stored by
// the game.)
message PlayerHoldings {
  required string name = 1;
  required uint32 money = 2;
  repeated Card hand = 3;
  repeated Card purchases = 4;
}

// The state for an auction (in progress or completed).
message Auction {
  // The one or two cards on auction.
  repeated Card cards = 1;

  // The name of the Player selling. (For an open double, of the two Players
  // jointly selling; the Player who played the open is first.)
  repeated string seller_names = 2;

  // True if putting up a card ended the round (and the auction should not
  // actually be performed).
  optional bool ends_round = 3;

  // Current high bid. If the winner is known, this is the winning bid.
  optional uint32 winning_bid = 4;

  // The name of the Player who won the auction. If omitted, the auction is
  // not yet complete.
  optional string winner_name = 5;
}

// Overall state of the game: common features for current action (the current
// auction, the deck) and past action (placings/values for each artist for
// previous rounds); and the official record of what each Player has (cards,
// money, etc).
message Board {
  repeated Card deck = 1;

  optional Auction auction = 2;

  message ArtistOutcome {
    required Artist.Id artist = 1;
    required uint32 value = 2;
  }
  message RoundOutcome {
    repeated ArtistOutcome artist_outcomes = 1;
  }
  repeated RoundOutcome round_outcomes = 3;

  repeated PlayerHoldings player_holdings = 4;
}
